cpp/src/sort/sort.cu

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/sorting.hpp>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/table/table_view.hpp>
+#include <utilities/error_utils.hpp>
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <thrust/sequence.h>
+
+namespace cudf {
+namespace exp {
+
+namespace detail {
+
+// Create permuted row indices that would materialize sorted order
+std::unique_ptr<column> sorted_order(table_view input,
+                                     std::vector<order> const& column_order,
+                                     null_order null_precedence,
+                                     cudaStream_t stream = 0) {
+  if (input.num_rows() == 0 or input.num_columns() == 0) {
+    return cudf::make_numeric_column(data_type{INT32}, 0);
+  }
+
+  if (not column_order.empty()) {
+    CUDF_EXPECTS(
+        static_cast<std::size_t>(input.num_columns()) == column_order.size(),
+        "Mismatch between number of columns and column order.");
+  }
+
+  std::unique_ptr<column> sorted_indices = cudf::make_numeric_column(
+      data_type{INT32}, input.num_rows(), mask_state::UNALLOCATED, stream);
+
+  mutable_column_view mutable_indices_view = sorted_indices->mutable_view();
+
+  auto device_table = table_device_view::create(input, stream);
+
+  thrust::sequence(rmm::exec_policy(stream)->on(stream),
```

Comment:
You could avoid having to write 0 to N to global memory here by using a counting iterator as the input in your sort lower down. You still need to allocate the space since its where you store the output. Its a very small optimization but a relatively easy one so I thought I would mention it.

Reply:
Normally I'd agree with you, but that's not possible in this case. `thrust::sort` doesn't have separate input and output iterators, as it's expected to be in-place. As such, if you were to pass counting iterators into a `thrust::sort`, you wouldn't materialize any output.

Reply:
What about cub::sort?

Reply:
CUB only has a radix sort. We need a comparison sort so we can plug in a custom row lexicographic comparator.

Reply:
Yup mea culpa. I forgot thrust only sorts in place